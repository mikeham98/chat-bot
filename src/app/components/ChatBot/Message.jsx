import React from 'react';
import PropTypes from 'prop-types';
import styles from '../../../themes/components/messages.scss';
import isPopulatedArray from "../../util/isPopulatedArray";
import Media from "./Media/Media";
import Options from "./Options/Options";
import {MediaObjectPropTypes, OptionsPropTypes} from "../../config/propTypes";

export default class Message extends React.PureComponent {
    constructor(props) {
        super(props);
        this.onClickOption = this.onClickOption.bind(this);
    }

    onClickOption(optionId) {
        const {id} = this.props;
        this.props.onClickOption(optionId, id)
    }

    showMedia() {
        return isPopulatedArray(this.props.media);
    }

    showOptions() {
        return isPopulatedArray(this.props.options);
    }

    render() {
        const {body, showDateTime, dateTime, media, options, onClickBody, selectedOption, color} = this.props;
        return (
            <div className={styles.messageContent}>
                <div className={styles.message} style={{backgroundColor: color}} onClick={onClickBody}>
                    <span className={styles.messageBody}>{body}</span>
                    {this.showOptions() && (
                        <Options
                            selectedOption={selectedOption}
                            options={options}
                            onClick={this.onClickOption}
                        />
                    )}
                </div>
                {this.showMedia() && (
                    <Media
                        media={media}
                    />
                )}
                {showDateTime && <span className={styles.messageCreatedAt}>{dateTime}</span>}
            </div>
        );
    }
}

Message.defaultProps = {
    color: ''
};

Message.propTypes = {
    onClickOption: PropTypes.func.isRequired,
    id: PropTypes.number,
    body: PropTypes.string,
    showDateTime: PropTypes.bool,
    dateTime: PropTypes.string,
    media: MediaObjectPropTypes,
    options: OptionsPropTypes,
    onClickBody: PropTypes.func.isRequired,
    selectedOption: PropTypes.number,
    color: PropTypes.string,
};