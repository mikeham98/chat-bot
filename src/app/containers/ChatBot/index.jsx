import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {getMessages, setOption} from "../../actions/chatBot/messages/index.actions";
import {sendMessage} from "../../actions/chatBot/toolbar/index.actions";
import ChatBot from "../../components/ChatBot";
import {chatBotSelector} from "../../selectors/chatBotSelector";
import {ConversationPropTypes, MessagesPropTypes} from "../../config/propTypes";

export class ChatBotContainer extends React.Component {
    constructor(props) {
        super(props);
        this.getMessages = this.getMessages.bind(this);
        this.sendMessage = this.sendMessage.bind(this);
        this.setOption = this.setOption.bind(this);
    }

    componentDidMount() {
        this.getMessages();
    }

    componentDidUpdate(prevProps) {
        if (this.props.currentConversation.id !== prevProps.currentConversation.id) {
            this.getMessages();
        }
    }

    getMessages() {
        this.props.getMessages(this.props.currentConversation.id);
    }

    sendMessage(message, callback) {
        const {id} = this.props.currentConversation;
        this.props.sendMessage(id, message, callback);
    }

    setOption(optionId, messageId) {
        const {id} = this.props.currentConversation;
        this.props.setOption(optionId, messageId, this.getMessages, id)
    }

    render() {
        const {messages, replying, currentUserId} = this.props;
        const {profile, color} = this.props.currentConversation;
        return (
            <ChatBot
                messages={messages}
                color={color}
                botName={profile && profile.name || ""}
                replying={replying}
                currentUserId={currentUserId}
                sendMessage={this.sendMessage}
                onClickOption={this.setOption}
            />
        )
    }
}

ChatBotContainer.propTypes = {
    getMessages: PropTypes.func.isRequired,
    sendMessage: PropTypes.func.isRequired,
    setOption: PropTypes.func.isRequired,
    messages: MessagesPropTypes,
    replying: PropTypes.bool,
    currentUserId: PropTypes.number,
    currentConversation: ConversationPropTypes,
};

const mapStateToProps = (state) => chatBotSelector(state);

const mapDispatchToProps = (dispatch) => {
    return {
        getMessages: (conversationId) => {
            dispatch(getMessages(conversationId))
        },
        sendMessage: (conversationId, message, callback) => {
            dispatch(sendMessage(conversationId, message, callback))
        },
        setOption: (optionId, messageId, callback, conversationId) => {
            // conversation id is passed in so that the application can get the next message on select of the option
            dispatch(setOption(optionId, messageId, callback, conversationId));
        }
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(ChatBotContainer);