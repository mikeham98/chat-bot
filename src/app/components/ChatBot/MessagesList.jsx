import React from 'react';
import classnames from 'classnames';
import styles from '../../../themes/components/messages.scss';
import isPopulatedArray from "../../util/isPopulatedArray";
import Message from "./Message";
import {left, right} from "../../config/messages.config";
import formatDateTime from "../../util/formatDateTime";

export default class MessagesList extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            clickedMessageId: null
        };
        this.messageContainerRef = this.messageContainerRef.bind(this);
    }

    componentDidMount() {
        this.scrollToBottom();
    }

    componentDidUpdate(prevProps) {
        if ((this.props.replying && !prevProps.replying) || (this.props.messages.length > prevProps.messages.length)) {
            this.scrollToBottom();
        }
    }

    scrollToBottom() {
        this.messageContainer.scrollTop = this.messageContainer.scrollHeight;
    }

    returnMessageClassName(position) {
        const classNames = [styles.messageWrapper];
        if (position === right) {
            classNames.push(styles.rightMessage);
        } else {
            classNames.push(styles.leftMessage);
        }
        return classnames(classNames);
    }

    returnMessages() {
        const {messages, currentUserId} = this.props;
        if (isPopulatedArray(messages)) {
            console.log(messages);
            return messages.map(message => {
                const position = currentUserId === message.createdBy.userId ? right : left;
                // showDateTime is implemented so that only one message at a time can show the date/time
                const showDateTime = message.id === this.state.clickedMessageId;
                return (
                    <div key={message.id}
                         onClick={() => this.setState({clickedMessageId: showDateTime ? null : message.id})}
                         className={this.returnMessageClassName(position)}>
                        <Message
                            showDateTime={showDateTime}
                            dateTime={formatDateTime(message.createdAt)}
                            body={message.content.body}
                        />
                    </div>
                );
            })
        }
    }

    messageContainerRef(ref) {
        this.messageContainer = ref;
    }

    render() {
        const {replying, botName} = this.props;
        return (
            <div
                ref={this.messageContainerRef}
                className={styles.messageContainer}
            >
                {this.returnMessages()}
                {replying && <span className={styles.replyInProgress}>{botName} is replying...</span>}
            </div>
        )
    }
}