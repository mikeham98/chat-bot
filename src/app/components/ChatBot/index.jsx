import React from 'react';
import PropTypes from "prop-types";
import styles from '../../../themes/components/chatBot.scss'
import MessagesList from "./MessagesList";
import MessageForm from "./Toolbar/MessageForm";
import {MessagesPropTypes} from "../../config/propTypes";

const ChatBot = ({messages, replying, currentUserId, color, botName, sendMessage, onClickOption}) => {
    return (
        <div className={styles.chatBot}>
            <MessagesList
                messages={messages}
                replying={replying}
                color={color}
                botName={botName}
                currentUserId={currentUserId}
                onClickOption={onClickOption}
            />
            <MessageForm
                sendMessage={sendMessage}
            />
        </div>
    );
};

ChatBot.defaultProps = {
    botName: ''
};

ChatBot.propTypes = {
    messages: MessagesPropTypes,
    replying: PropTypes.bool,
    currentUserId: PropTypes.number,
    color: PropTypes.string,
    botName: PropTypes.string,
    sendMessage: PropTypes.func.isRequired,
    onClickOption: PropTypes.func.isRequired,
};

export default React.memo(ChatBot);