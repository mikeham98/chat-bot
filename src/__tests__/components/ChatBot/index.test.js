import React from 'react';
import {shallow} from 'enzyme';
import ChatBot from '../../../app/components/ChatBot/index';

describe('ChatBot', () => {

    let shallowedWrapper;
    let props = {};
    const wrapper = () => {
        if (!shallowedWrapper) {
            shallowedWrapper = shallow(<ChatBot {...props}/>);
        }
        return shallowedWrapper;
    };

    beforeEach(() => {
        props = {
            messages: [1,2,3,4,5],
            replying: true,
            currentUserId: 1,
            botName: 'bot1',
            sendMessage: jest.fn(),
            onClickOption: jest.fn()
        };
        shallowedWrapper = undefined;
    });

    describe('render', () => {
        it('should return MessageList and MessageForm', () => {
            expect(wrapper()).toMatchSnapshot();
        });
    });

    describe('functionality', () => {
        describe('prop', () => {
            describe('onClickOption', () => {
                it('should be called on onClick of message div', () => {
                    expect(props.onClickOption).not.toHaveBeenCalled();
                    wrapper().find('MessagesList').first().props().onClickOption();
                    expect(props.onClickOption).toHaveBeenCalledTimes(1);
                });
            });
            describe('sendMessage', () => {
                it('should be called on onClick of Options component', () => {
                    expect(props.sendMessage).not.toHaveBeenCalled();
                    wrapper().find('MessageForm').first().props().sendMessage();
                    expect(props.sendMessage).toHaveBeenCalledTimes(1);
                });
            });
        });
    });
});