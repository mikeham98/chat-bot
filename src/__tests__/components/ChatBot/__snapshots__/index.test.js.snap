// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBot render should return MessageList and MessageForm 1`] = `
<div
  className="chatBot"
>
  <MessagesList
    botName="bot1"
    currentUserId={1}
    messages={
      Array [
        1,
        2,
        3,
        4,
        5,
      ]
    }
    onClickOption={[MockFunction]}
    replying={true}
  />
  <MessageForm
    initialValues={
      Object {
        "message": "",
      }
    }
    sendMessage={[MockFunction]}
  />
</div>
`;
